### PR TITLE
Support ECDSA(secp256k1) signatures

### LINKED ISSUE #2428: Add a `JKey` specialization for ECDSA(secp256k1) keys
⚠️ Blocked by https://github.com/hashgraph/hedera-protobufs/issues/110

---

### Problem

To support ECDSA secp256k1 keys, we need a specialization of the abstract `JKey` type that supports secp256k1, say `JECDSAsecp256k1Key`.

### Solution

Update the `JKey` type hierarchy to:

- [x] Add a `JECDSAsecp256k1Key` type whose `isValid()` implementation requires a source `byte[]` of length 33 beginning with exactly `0x02` or `0x03`. (And if necessary for crypto performance, internally maintains both the compressed and uncompressed representations.)
- [x]  Update the gRPC-to-`JKey` ["mapping factory"](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JKey.java#L111) to translate a gRPC `Key` object with non-empty `bytes ecdsaSecp256k1` into a `JECDSAsecp256k1Key` instance.
- [x] Update the `JKey`-to-gRPC ["view factory"](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JKey.java#L223) to translate a `JECDSAsecp256k1Key` instance to a `Key` whose `oneof` has non-empty `bytes ecdsaSecp256k1`.
- [x] Update the `JKey` serdes (`serialize()` + `deserialize()`) to support this new key type.
- [x] Unit test coverage

### LINKED ISSUE #2429: Update `PlatformSigOps` to extract (alleged) signatures for both Ed25519 and ECDSA secp256k1 keys
:warning: Along with #2428, blocked pending SDK support for ECDSA secp256k1 signature verification in the `Cryptography` module.

---

### Problem

Even given protobuf and `JKey` support for ECDSA secp256k1, and an SDK that supports ` SignatureType.ECDSA_SECP256K1`; the current `expandSignatures()` code path will not expand appropriate `TransactionSignature` instances that could be used to activate an account's secp256k1 keys. 

The issue is that [the `PlatformSigOps.createEd25519PlatformSigsFrom()` method](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/sigs/PlatformSigOps.java#L54) assumes any "simple" `JKey` represents an Ed25519 public key. (Clue in the name!  :mushroom:)

### Solution

Generalize to a `PlatformSigOps.createPlatformSigsFrom(...)` method that can create a `TransactionSignature` instance appropriate for **either** a Ed25519 or ECDSA(secp256k1) key.

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
- Enable use of ECDSA(secp256k1) keys wherever Ed25519 keys are supported.

**Related issues**:
- Closes #2428 
- Closes #2429

### CODE REVIEW

File: hedera-node/pom.xml [hunk 1]
```diff
@@ -257,6 +258,10 @@
       <groupId>com.swirlds</groupId>
       <artifactId>swirlds-fchashmap</artifactId>
     </dependency>
+    <dependency>
+      <groupId>com.swirlds</groupId>
+      <artifactId>swirlds-jasperdb</artifactId>
+    </dependency>
```
Comment (Neeharika-Sompalli) [inline hedera-node/pom.xml:264]: do we need these dependencies  without VM ?

File: hedera-node/pom.xml [hunk 2]
```diff
@@ -265,6 +270,10 @@
       <groupId>com.swirlds</groupId>
       <artifactId>swirlds-fcqueue</artifactId>
     </dependency>
+    <dependency>
+      <groupId>com.swirlds</groupId>
+      <artifactId>swirlds-virtualmap</artifactId>
+    </dependency>
     <dependency>
```
Comment (Neeharika-Sompalli) [inline hedera-node/pom.xml:277]: same here

File: hedera-node/pom.xml [hunk 3]
```diff
@@ -315,6 +324,11 @@
       <artifactId>besu-datatypes</artifactId>
       <version>${besu.version}</version>
     </dependency>
+    <dependency>
+      <groupId>com.github.ben-manes.caffeine</groupId>
+      <artifactId>caffeine</artifactId>
+      <version>${caffeine.version}</version>
+    </dependency>
   </dependencies>
```
Comment (Neeharika-Sompalli) [inline hedera-node/pom.xml:332]: same here

File: hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JDelegatableContractIDKey.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package com.hedera.services.legacy.core.jproto;
+
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JDelegatableContractIDKey.java:2]: This class can be removed?

File: hedera-node/src/main/java/com/hedera/services/sigs/sourcing/PubKeyToSigBytes.java [hunk 5]
```diff
@@ -44,4 +46,38 @@
...
+	 * instances based on {@link com.hederahashgraph.api.proto.java.SignaturePair}s
+	 * with a full public key prefix, but <b>without</b> a matching public
+	 * key obviously linked to the parent transaction.
+	 *
+	 * A canonical example would the the supply key of a token that will be
+	 * minted through an HTS precompile run as part of a {@code ContractCall}
+	 * transaction. This key is not <b>obviously</b> linked to the top-level
+	 * {@code ContractCall}; but it will ultimately need to sign, hence requires
+	 * a signature supplied with full public key prefix.
+	 *
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/sigs/sourcing/PubKeyToSigBytes.java:66]: ❤️

File: hedera-node/src/main/java/com/hedera/services/sigs/Rationalization.java [hunk 6]
```diff
@@ -121,6 +122,10 @@ private void execute() {
 			otherFailure = otherPartiesStatus;
 		} else {
 			reqOthersSigs = lastOrderResult.getOrderedKeys();
+			if (pkToSigFn.hasAtLeastOneUnusedSigWithFullPrefix()) {
+				pkToSigFn.forEachUnusedSigWithFullPrefix((pubKey, sig) ->
+						realOtherPartySigs.add(bodySigningFactory.signBodyWithEd25519(pubKey, sig)));
```
Comment (anighanta) [inline hedera-node/src/main/java/com/hedera/services/sigs/Rationalization.java:127]: need support for `secp256k1` key as well here

File: hedera-node/src/main/java/com/hedera/services/sigs/Expansion.java [hunk 7]
```diff
@@ -78,20 +78,27 @@ public ResponseCodeEnum execute() {
...
 						otherStatus);
 			}
+			return otherStatus;
 		}
-		return otherStatus;
+
+		if (pkToSigFn.hasAtLeastOneUnusedSigWithFullPrefix()) {
+			pkToSigFn.forEachUnusedSigWithFullPrefix((pubKey, sig) ->
+					txnAccessor.getPlatformTxn().add(sigFactory.signBodyWithEd25519(pubKey, sig)));
+		}
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/sigs/Expansion.java:87]: here as well ?

File: pom.xml [hunk 8]
```diff
@@ -415,11 +415,21 @@ ${git.signoff}
         <artifactId>swirlds-fcqueue</artifactId>
         <version>${swirlds.version}</version>
       </dependency>
+      <dependency>
+        <groupId>com.swirlds</groupId>
+        <artifactId>swirlds-jasperdb</artifactId>
+        <version>${swirlds.version}</version>
+      </dependency>
```
Comment (anighanta) [inline pom.xml:422]: same here .. we dont require these yet.

Comment (Neeharika-Sompalli) [reply pom.xml]: removed the VM and jasperDB dependecies

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM !

**Review by anighanta — APPROVED** [review]

**Pull request merged**